### PR TITLE
Adding notebook and one-hot encoding functions

### PR DESCRIPTION
**Pull request by Alissa13777**
Notebook with tests on 15 files stored locally and plotting of results : one-hot_encoding.ipynb
Functions of one-hot encoding : hot_encoding_utils.py

### CODE REVIEW

File: asip_v2/hot_encoding_utils.py [hunk 1]
```diff
@@ -0,0 +1,137 @@
...
+    Returns
+    -------
+    index_ : integer
+        index of the list where the value (0/1 or concentration) will be.
+
+    """
+    index_=0
+    #if stage ==0:
+        #index_ = 0
+        #print('ice_free')
```
Comment (akorosov) [inline asip_v2/hot_encoding_utils.py:32]:
```suggestion
    index_= np.nan
    if stage ==0:
        index_ = 0
        #print('ice_free')

```

File: asip_v2/hot_encoding_utils.py [hunk 2]
```diff
@@ -0,0 +1,137 @@
...
+    if stage!=-9:
```
Comment (akorosov) [inline asip_v2/hot_encoding_utils.py:33]:
```suggestion

```

File: asip_v2/hot_encoding_utils.py [hunk 3]
```diff
@@ -0,0 +1,137 @@
...
+        if stage in range(86,94):
+            #print('First year ice')
+            index_=2
+        if stage in range(95,98):
+            #print('multiyear ice')
+            index_=3
+    return index_
+    
+
+def one_hot_m1(ct,ca,sa,fa,cb,sb,fb,cc,sc):
```
Comment (akorosov) [inline asip_v2/hot_encoding_utils.py:46]:
```suggestion
def one_hot_m1(ct,ca,sa,fa,cb,sb,fb,cc,sc,fc, min_ct=10):

```

File: asip_v2/hot_encoding_utils.py [hunk 4]
```diff
@@ -0,0 +1,137 @@
...
+    if index ==0:
+        result = [1,0,0,0]
+    else :
+        #print([sa,sb,sc])
+        #print('bibi',[sa,sb,sc][index])
+        index2 = form_of_ice([sa,sb,sc][index])
+        #print(index2)
+        result = [0,0,0,0]
+        result[index2]=1
+    return result
```
Comment (akorosov) [inline asip_v2/hot_encoding_utils.py:93]:
```suggestion
    L=[ct,ca,cb,cc]
    result = [0,0,0,0]
    if ct < min_ct:
        index == 0
    else :
        index = np.argmax(L[1:])
        index = form_of_ice([sa,sb,sc][index])
    result[index]=1
    return result

```

**Review by akorosov — CHANGES_REQUESTED** [review]

File: asip_v2/hot_encoding_utils.py [hunk 5]
```diff
@@ -0,0 +1,135 @@
...
+    stage : integer
+        stage of development.
+
+    Returns
+    -------
+    index_ : integer
+        index of the list where the value (0/1 or concentration) will be.
+
+    """
+    index_= np.nan
```
Comment (akorosov) [inline asip_v2/hot_encoding_utils.py:29]:
```suggestion
    index_= None

```

File: asip_v2/hot_encoding_utils.py [hunk 6]
```diff
@@ -0,0 +1,135 @@
...
+    """
+    
+    L=[ct,ca,cb,cc]
+    result = [0,0,0,0]
+    if ct < min_ct:
+        index == 0
+    else :
+        index = np.argmax(L[1:])
+        index = form_of_ice([sa,sb,sc][index])
+    result[index]=1
```
Comment (akorosov) [inline asip_v2/hot_encoding_utils.py:89]:
```suggestion
    cabc = [ca,cb,cc]
    result = [0,0,0,0]
    if ct < min_ct:
        return [1,0,0,0] # open water
        
    f = [0,0,0] # fractions
    for ci,si in zip([ca,cb,cc], [sa,sb,sc]):
        icetype = ice_type(si)
        if ci != -9 and icetype is not None:
           f[sod-1] += ci
        if max(f) == 0:
            icetype = ice_type(sa)
        else:
            icetype = np.argmax(f)+1
    result[icetype]=1

```

**Review by akorosov — CHANGES_REQUESTED** [review]

File: asip_v2/archive.py [hunk 7]
```diff
@@ -1,442 +1,471 @@
...
-                self.PROP.update(obj.calculate_variable_ML())
-                del obj
-            del self.mask_batches_amsr2, self.mask_batches
+import os
+from json import dump, load
+
+import numpy as np
+from skimage.util.shape import view_as_windows
+from scipy.ndimage import uniform_filter
+from hot_encoding_utils import one_hot_m1, one_hot_m2, ice_type
```
Comment (Alissa13777) [inline asip_v2/archive.py:7]: changed

Comment (Alissa13777) [inline asip_v2/archive.py:7]:
```suggestion
from hot_encoding_utils import one_hot_m1, one_hot_m2, ice_type

```

File: asip_v2/archive.py [hunk 8]
```diff
@@ -1,442 +1,471 @@
...
+    classes.
+    """
+    def view_as_windows(self, array):
+        print("array.shape",len(array.shape))
+        if len(array.shape)==3 and len(self.WINDOW_SIZE)==2:
+            n,p = self.WINDOW_SIZE
+            q = array.shape[2]
+            self.WINDOW_SIZE = (n,p,q)
+        print("self.WINDOW_SIZE", self.WINDOW_SIZE)
+        return view_as_windows(array, self.WINDOW_SIZE, self.STRIDE)
```
Comment (Alissa13777) [inline asip_v2/archive.py:21]:
```suggestion
    def view_as_windows(self, array):
        print("array.shape",len(array.shape))
        if len(array.shape)==3 and len(self.WINDOW_SIZE)==2:
            n,p = self.WINDOW_SIZE
            q = array.shape[2]
            self.WINDOW_SIZE = (n,p,q)
        print("self.WINDOW_SIZE", self.WINDOW_SIZE)
        return view_as_windows(array, self.WINDOW_SIZE, self.STRIDE)

```

File: asip_v2/archive.py [hunk 9]
```diff
@@ -1,442 +1,471 @@
...
+        
+         # oroginal dictionary should not be changed
+        for id_value, variable_belong_to_id in self.map_id_to_variable_values.items():
+        # each loop changes all locations of values_array (that have the very
+        # 'id_value') to its corresponding value inside 'variable_belong_to_id'
+               
+            #Filling the 3D array
+            en_values_array[ic == id_value, :] = np.byte(np.array(variable_belong_to_id))
+       
+        return en_values_array
```
Comment (Alissa13777) [inline asip_v2/archive.py:166]:
```suggestion
    def encode_icechart(self, values_array, element):
        """
        based on 'self.map_id_to_variable_values', all the values are converted to correct values
        of the very variable based on polygon ID values in each location in 2d array of values_array
        """
        # print(values_array.shape)
        ic = values_array.copy()
        n,p= ic.shape
        #construction of the 3D array that will be filled
        en_values_array=np.zeros((n,p,4))+np.nan
        
         # oroginal dictionary should not be changed
        for id_value, variable_belong_to_id in self.map_id_to_variable_values.items():
        # each loop changes all locations of values_array (that have the very
        # 'id_value') to its corresponding value inside 'variable_belong_to_id'
               
            #Filling the 3D array
            en_values_array[ic == id_value, :] = np.byte(np.array(variable_belong_to_id))
       
        return en_values_array

```

File: asip_v2/archive.py [hunk 10]
```diff
@@ -1,442 +1,471 @@
...
+            result = one_hot_m1(ct,ca,sa,fa,cb,sb,fb,cc,sc,fc)
+            ## if you want the method with the concentration uncomment the following line 
+    #         result = one_hot_m2(ct,ca,sa,fa,cb,sb,fb,cc,sc)
+    
+            #Filling the dictionnary
+            map_id_to_variable_values.update({int(id_val_splitted[0]): result})
+            #print("----- input \n",ct,ca,sa,fa,cb,sb,fb,cc,sc, "\n-----output \n",result,"\n")
+            #print("----- input \n",id_val_splitted, "\n-----output \n",result,"\n")
+
+        self.map_id_to_variable_values = map_id_to_variable_values
```
Comment (Alissa13777) [inline asip_v2/archive.py:309]:
```suggestion
    def read_icechart_coding(self, fil, filename):
        """
        based on 'polygon_codes' and 'polygon_icechart' section of netCDF file as well as the name
        of the file, this function set the values of properties of 'polygon_ids', 'scene',
        'names_polygon_codes' and 'map_id_to_variable_values' of archive object.
        """
        self.scene = filename.split('_')[0]
        # just from beginning up to variable 'FC' is considered, thus it is [:11] in line below
        self.names_polygon_codes = fil['polygon_codes'][0].split(";")[:11]
        self.polygon_ids = np.ma.getdata(fil["polygon_icechart"])
        map_id_to_variable_values = {}  # initialization
        # this dictionary has the ID as key and the corresponding values
        # as a list at the 'value postion' of that key in the dictionary.
        for id_and_corresponding_variable_values in fil['polygon_codes'][1:]:
            id_val_splitted = id_and_corresponding_variable_values.split(";")
            #print(id_val_splitted)
            [ct, ca, sa, fa, cb, sb, fb, cc, sc, fc] = list(map(int, id_val_splitted[1:11]))
            #result of the one-hot encoding method 
            result = one_hot_m1(ct,ca,sa,fa,cb,sb,fb,cc,sc,fc)
            ## if you want the method with the concentration uncomment the following line 
    #         result = one_hot_m2(ct,ca,sa,fa,cb,sb,fb,cc,sc)
    
            #Filling the dictionnary
            map_id_to_variable_values.update({int(id_val_splitted[0]): result})
            #print("----- input \n",ct,ca,sa,fa,cb,sb,fb,cc,sc, "\n-----output \n",result,"\n")
            #print("----- input \n",id_val_splitted, "\n-----output \n",result,"\n")

        self.map_id_to_variable_values = map_id_to_variable_values

```

**Pull request closed without merging**